### ISSUE TITLE
Implement ops needed for object detection models

### ISSUE THREAD

**Issue by ethanluoyc** [google/jax#16021]
- [x] Check for duplicate requests.
- [x] Describe your goal, and if possible provide a code snippet with a motivating example.

I am interested in porting some object detection models from PyTorch to JAX. In particular, I am looking at the RCNN family of models. 

After looking into some existing implementations in TF/PyTorch, I realize that there is currently not a straightforward way to port these models to JAX. In particular, the RCNN models require an ROIPool / ROIAlign ops, but I have not yet found a native way to do this in JAX. Having these ops would enable us to implement some popular object detection models such as those from [Detectron2](https://github.com/facebookresearch/detectron2).

I would like to know from the JAX developer's point of view what the best strategy to do this in JAX.

Here are some additional information regarding these ops.

## ROIAlign
I found a Python implementation of ROIAlign here https://github.com/pytorch/vision/blob/296bca682c8fd7fc9b91c0f38e0a475ec68b0863/torchvision/ops/roi_align.py and I have managed to port this to JAX. Otherwise, the usual way to do this in PyTorch uses a native op [roi_align](https://pytorch.org/vision/main/generated/torchvision.ops.roi_align.html)

I also found some TF implementation that uses tf.image.crop_and_resize but I can't find a JAX equivalent in `jax.image`. Also, an implementation using `crop_and_resize` as the `crop_and_resize` does not have a native op in XLA.

## ROIPool

I couldn't find a Python implementation of this anywhere and this usually requires a native op implementation. ROIPool requires cropping ROI part of the input image, this produces a dynamically-shaped array that cannot be JIT compiled. For PyTorch, they have a native op `roi_pool` specifically for this.

## More general questions
I think writing custom ops for these would be a solution. For that, I have some general questions.
1. If I implement the custom ops for these (CPU/GPU), it means that we won't be able to run these ops on TPUs, is that correct?
2. In TF, there is some support for dispatching a CPU kernel via a thread pool for improved parallelism, is there any similar machinery in JAX that would enable us to do similar things?

**Comment by froystig** [comment]
Have you looked into the [Scenic](https://github.com/google-research/scenic) library, and does it offer any of what you're after? A quick repository search there reveals an internal function called [`roi_align`](https://github.com/google-research/scenic/blob/1259d183c934f63610f325e5f066714bb65b83e4/scenic/projects/owl_vit/layers.py#L591) that bottoms out in `jax.image.scale_and_translate`.

> [...] this produces a dynamically-shaped array that cannot be JIT compiled.

It would be useful to have a minimal code snippet describing this function.

cc @sharadmv for escape hatches and custom kernels.

**Comment by froystig** [comment]
> If I implement the custom ops for these (CPU/GPU), it means that we won't be able to run these ops on TPUs, is that correct?

What mechanism do you have in mind for "custom ops" here?

**Comment by ethanluoyc** [comment]
> Have you looked into the [Scenic](https://github.com/google-research/scenic) library, and does it offer any of what you're after? A quick repository search there reveals an internal function called [`roi_align`](https://github.com/google-research/scenic/blob/1259d183c934f63610f325e5f066714bb65b83e4/scenic/projects/owl_vit/layers.py#L591) that bottoms out in `jax.image.scale_and_translate`.
> 
> > [...] this produces a dynamically-shaped array that cannot be JIT compiled.
> 
> It would be useful to have a minimal code snippet describing this function.
> 
> cc @sharadmv for escape hatches and custom kernels.

Yes indeed I have checked out Scenic and they seem to have roi_align, but I wasn't able to find out how to use it so that its output matches the behavior from PyTorch or if it's behaving entirely differently. I experimented with some example inputs but it's giving me output that's different from PyTorch. However, for `roi_align`, I do manage to get a working pure JAX version based on https://github.com/pytorch/vision/blob/296bca682c8fd7fc9b91c0f38e0a475ec68b0863/torchvision/ops/roi_align.py

For `roi_pool`, I haven't managed to figure out how to do this. I don't have an example for how to do this in pure numpy atm but maybe it would be useful to have an example so here it is.

Consider function `roi_pool(input, roi, output_shape)` where `input` is a 2D image
```
0  1  2   3 4
5  6  7   8 9
10 11 12 13 14
15 16 17 18 19
20 21 22 23 24
```
and `roi` is the Region of interest (bounding box) with the format (x0, y0, x1, y1) where (x0, y0) and (x1, y1) denotes the top-left and bottom-right of the bounding box respectively. Suppose that `roi = (1, 1, 3, 3)`. `output_shape` is the shape of the output after pooling, say we use `(3,3)`. Then we can do roi_pool in two steps.

1. We first crop out the ROI, in this case it would be
```
[ 6.,  7.,  8.],
[11., 12., 13.]
[16., 17., 18.]
```
2. we then max pool the cropped ROI to have shape of `output_shape`. In this case the cropped shape is the same as output so the result of pooling is the same.

What I thought would be difficult is that in step 1 the size of the cropped depends on concrete coordinates of the `roi` so I can not use sth like `jax.lax.dynamic_slice` to perform that as this intermediate step produces dynamically-shaped arrays (worse, even if JAX supports dynamic shape this still cannot be expressed in pure JAX because if we consider a batched op then each crop in the batch would have different shape, I think that might be the reason why in TF you have `tf.image.crop_and_resize`)

> What mechanism do you have in mind for "custom ops" here?
This is more of an arbitrary comment as I am trying to understand how JAX primates work on TPUs. I am learning about writing custom ops for JAX based on the official doc and also the https://github.com/dfm/extending-jax. I think in this case I can write the CPU/GPU custom ops and create the corresponding bindings.
